### ISSUE TITLE
Add Savitzky-Golay smoothing and derivatives as a preprocessing option

### ISSUE THREAD

**Issue by bitjungle** [bitjungle/gopca#843]
GoPCA has no derivative preprocessing. For spectroscopy that is the largest remaining gap in the preprocessing set: after SNV, taking a first or second derivative is the most common next step in an NIR workflow, because it removes additive baseline offset (1st derivative) and baseline slope (2nd derivative) that scatter correction alone leaves behind.

A **Savitzky-Golay** filter is the standard way to do it: fit a low-order polynomial across a sliding window of wavelengths and evaluate the polynomial or its derivative at the centre. It smooths and differentiates in one pass, which matters because differentiating noisy spectra directly amplifies the noise.

## Measured evidence, on a dataset we ship

`testdata/bronir2/bronir2.csv` — 441 labelled NIR spectra, 1001 wavelengths (1300–2300 nm), 13 products. Predicting `Dens#target`, validated **leave-one-product-out** so the estimate reflects performance on a product the model has not seen:

| Preprocessing | Q² at fixed k=15 | Best Q², k ∈ 1..20 |
|---|---|---|
| Raw | 0.403 | 0.659 (k=20) |
| SNV, full range | — | 0.734 (k=20) |
| **SNV + 1st derivative, 1300–2100 nm** | **0.760** | **0.788** (k=19) |

The fixed-k row is the more striking comparison and the more relevant one: derivative preprocessing reaches at k=15 what the raw spectra need far more components to approach, and at k=20 never quite reach. Fewer components for better prediction is exactly what good preprocessing buys.

Two honest caveats. The window length (11) and polynomial order (2) are common defaults, not tuned — a real implementation should expose both. And comparing preprocessing choices by cross-validated Q² makes that Q² optimistic as a performance estimate; these numbers are evidence that the transform helps, not a claim about achievable accuracy.

<details>
<summary>Reproduction script (needs <code>testdata/.venv</code>)</summary>

```python
import numpy as np, pandas as pd
from scipy.signal import savgol_filter
from sklearn.decomposition import PCA
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import LeaveOneGroupOut

df = pd.read_csv("testdata/bronir2/bronir2.csv", index_col=0)
y_all = df["Dens#target"].to_numpy(float)
spec = [c for c in df.columns if "#" not in c and pd.api.types.is_numeric_dtype(df[c])]
wl = np.array([float(c) for c in spec]); X_all = df[spec].to_numpy(float)
lab = np.isfinite(y_all) & np.isfinite(X_all).all(axis=1)
X, y, g = X_all[lab], y_all[lab], df["Product"].astype(str).to_numpy()[lab]

snv = lambda A: (A - A.mean(1, keepdims=True)) / A.std(1, keepdims=True)
band = (wl >= 1300) & (wl <= 2100)

for name, A in [("raw", X),
                ("SNV+1st deriv", savgol_filter(snv(X)[:, band], 11, 2, deriv=1, axis=1))]:
    oof = np.empty_like(y)
    for tr, te in LeaveOneGroupOut().split(A, y, g):
        sc = StandardScaler().fit(A[tr]); p = PCA(n_components=15).fit(sc.transform(A[tr]))
        lr = LinearRegression().fit(p.transform(sc.transform(A[tr])), y[tr])
        oof[te] = lr.predict(p.transform(sc.transform(A[te])))
    print(name, round(1 - np.sum((oof-y)**2)/np.sum((y-y.mean())**2), 3))
```

</details>

## Sketch of the work

- Implement in `internal/core/preprocessing.go` alongside SNV and vector normalisation. It is **row-wise** and operates along the variable axis, so it belongs in the same family and shares their ordering constraints (before column centring and scaling).
- Parameters: window length (odd, > polynomial order), polynomial order, derivative order (0 for smoothing only, 1, 2).
- Edge handling needs a decision. `scipy` defaults to `mode="interp"`, fitting the polynomial to the end window rather than padding; that is the behaviour spectroscopists expect and is worth matching deliberately rather than by accident.
- Like SNV, it makes `OriginalScaleValid` false for PCR — a derivative is a fixed linear operator on the wavelength axis, so unlike SNV it *does* have a closed-form collapse back to original variables, but only if the operator is recorded. Worth deciding explicitly rather than defaulting to "unavailable".
- Surface in the CLI (`--savgol-window`, `--savgol-order`, `--savgol-deriv` or similar), GoPCA Desktop preprocessing panel, and the model artifact + schema so `pca transform` can reapply it.
- Docs: `intro_to_pca.md` Section 5 (Advanced Preprocessing) and the Corn dataset section, which currently stops at SNV.

## Relationship to other issues

Companion to #774 (EMSC). Both are scatter/baseline corrections for spectroscopy and both are row-wise; EMSC is the more sophisticated and needs a reference spectrum, while Savitzky-Golay is simpler, parameter-light and more widely used. If only one is implemented, this is the one with broader reach.

Also relevant to #832: `intro_to_pca.md` Section 12 now advises spectroscopists to check their preprocessing against a cross-validated model rather than the scores plot alone. That advice is more useful when the preprocessing set includes derivatives.

**Comment by bitjungle** [comment]
All three parts are merged: the engine (#888), the CLI, model artifact and PCR collapse (#890), and GoPCA Desktop (#891). Every item in the original scope is delivered, including the docs in `intro_to_pca.md` §5 and `cli_reference.md`.

Two things worth recording that the original issue left open.

### The PCR question, answered

The issue asked whether the original-scale coefficient collapse should be available. It is, and the reasoning is now explicit in the code: Savitzky-Golay acts along a row, which is exactly what makes SNV uncollapsible — but it is the *same* operator for every sample, so it folds into the coefficients as a transpose and a deployed model needs the raw wavelengths and nothing else. Combined with SNV the collapse is correctly unavailable again, because SNV still scales each sample by its own spread.

### Why SNV and the derivative are complementary, not alternatives

This came up when the panel was reviewed, and it is worth having on the record because it is the first question the feature prompts. Differentiating removes any constant, so SNV's mean-subtraction vanishes:

```
savgol_deriv(SNV(x)) = savgol_deriv(x) / sd(x)
```

Verified against scipy on the corn spectra to 1.8e-14 (1st derivative) and 2.9e-14 (2nd). What survives SNV is the division by each sample's spectral spread — the **multiplicative** scatter correction a derivative cannot perform. Scattering both shifts the baseline and stretches the spectrum; the derivative handles the first and the divisor the second, which is why the pair beats either alone in this issue's own bronir2 measurement.

For smoothing (`--savgol-deriv 0`) nothing vanishes and SNV applies in full. `--vector-norm` behaves identically at every derivative order, since it never centres.

Follow-up work on making the Desktop spectra-aware — a live preview of the preprocessed spectra and a check that the variable axis is actually a continuum — is tracked separately.
